fastar/rules.py

```diff
@@ -250,3 +248,36 @@ def dot_general_backward_rule(outbox, lhs, rhs, dimension_numbers, precision):
   return [lhs_start, rhs_start], [lhs_incount, rhs_incount]
 
 def_standard_op(lax.dot_general_p, dot_general_backward_rule)
+
+def pad_backward_rule(outbox, operand, padding_value, padding_config):
+  out_start, out_shape = outbox
+  lo, _, interior = unzip3(padding_config)
+  dilation = np.array(interior) + 1
+  assert type(out_start) == np.ndarray
+  inclip = lambda indices: np.clip(indices, 0, operand.shape)
+  lo_sign = np.where(np.less(lo, 0), -1, 1)
+  instart = inclip(lo_sign * np.floor_divide(lo_sign * (out_start - lo), dilation))
+  instop = inclip(lax.lax._ceil_divide(out_start + out_shape - lo, dilation))
+  inshape = instop - instart
+  padcount = prod(out_shape) - prod(inshape)
+  return [instart, ()], [np.ones(inshape), padcount]
+
+def pad_update_rule(cache, outbox, operand, padding_value, padding_config):
+  out_start, out_shape = outbox
+  (instart, _), (incount, _) = pad_backward_rule(outbox, operand, padding_value, padding_config)
+  inslice = lax.dynamic_slice(operand, instart, incount.shape)
+  lo, _, interior = unzip3(padding_config)
+  dilation = np.array(interior) + 1
+  inclip = lambda indices: np.clip(indices, 0, operand.shape)
+  next_instart = inclip(lax.lax._ceil_divide(out_start - lo, dilation))
+  next_outstart = next_instart * dilation + lo
+  _start = next_outstart - out_start
+  _stop = _start + jnp.array(incount.shape) * dilation
+  out_index = [slice(start, stop, step)
+               for start, stop, step in zip(_start, _stop, dilation)]
+  out_slice = jnp.full(out_shape, padding_value)
+  out_slice = ops.index_update(out_slice, out_index, inslice)
```

Comment:
I'm just wondering whether we can express the creation of out_slice in terms of lax.pad, I'm having a go at it now.

Reply:
Totes, good idea!

Reply:
@JuliusKunze do you wanna do it actually? I spent 10 mins or so trying to do it last night but didn't get very far.

Reply:
Generally I think we should attempt to avoid anything which bottoms out on a gather, scatter or reshape, because they're known to be slow to compile, and therefore slow in op-by-op.

Reply:
Sure, will do

Reply:
Done!